src/core/net/netif.cpp

```diff
@@ -301,14 +365,20 @@ otError Netif::SubscribeExternalMulticast(const Address &aAddress)
 {
     otError                error = OT_ERROR_NONE;
     NetifMulticastAddress *entry;
+    NetifMulticastAddress &linkLocalAllRoutersAddress = static_cast<NetifMulticastAddress &>(
+        const_cast<otNetifMulticastAddress &>(kLinkLocalAllRoutersMulticastAddress));
 
-    VerifyOrExit(!mMulticastAddresses.IsEmpty(), error = OT_ERROR_INVALID_STATE);
+    // Check that the address is not one of the fixed addresses:
+    // LinkLocalAllRouters -> RealmLocalAllRouters -> LinkLocalAllNodes
+    // -> RealmLocalAllNodes -> RealmLocalAllMpl.
 
-    if (IsMulticastSubscribed(aAddress))
+    for (const NetifMulticastAddress *cur = &linkLocalAllRoutersAddress; cur; cur = cur->GetNext())
```

Comment:
Is it a must to check these addresses given that they will be checked in `IsMulticastSubscribed()`?
The linux kernel may subscribe ff02::2 on every interface so returning OT_ERROR_ALREADY may be more friendly.

Reply:
The main change in this PR is we allow user to subscribe to multicast addresses at any point even before these addresses are added by the OpenThread core code to interface (so when calls these API, these addresses they may not yet be in the list). Returning `OT_ERROR_ALREADY` won't be correct then since the addresses are not really subscribed.

Note that OT core will add 3 addresses (link local and realm local all nodes and all MPL forwarders)  when netif is brought up. The other two router specific ones (link local and realm local all routers) are added only when device becomes a router. 

This behavior from this method is similar to how we don't let user add a unicast link local address, i.e. `AddExternalUnicastAddress()` returning `OT_ERROR_INVALID_ARGS` if the given address is link-local.